### ISSUE TITLE
[Enhancement] Webpack and friends as alternative for underlying build/bundling system.

### ISSUE THREAD

**Issue by TheLarkInn** [angular/angular-cli#909]
🐘 Alright its time to address the elephant in the room. 🐘 

### Lets talk about Webpack integration
CLI Is incrediballz, but SystemJS is a less-than-desirable and less-than-configuration-friendly experience. 

### Use Webpack to have an awesome experience for all environments
CLI was built to make rapid development easy and awesome with Angular and I don't think it should stop there. Out of the box, CLI users should have a bundling system in place that automagically handles giving them a production ready bundle of their Angular code that they spent the tireless 20 command lines to boilerplate. 

### Things should be easy to configure
Additionally, handling external libraries, build configuration, async module loading, source mapping, code splitting, code sharing, beautifying, etc. should be EASY to reconfigure if a user wants to. Currently it isn't.

### Shouldn't replace SystemJS
I do not think that we should rid of SystemJS entirely, however I do believe that it is important that we provide an alternative for people to use. (SystemJS has provided very successful for plunks and apps no larger than 300kb)

### `ng init --use-webpack`
Adding a simple flag like `ng init --use-webpack` might be a tolerable alternative that people would be more than happy to use. 

### There are great repo's out there for a starting point
I think @gdi2290 @foxandxss 's starters are community wide accepted examples of multi environment webpack setups that would be perfect as a starting point. Plus when Weback 2 ships from beta, we'll be able to leverage tree-shaking, async module loading, etc.

### I can probably help too
I will add to this issue a more concrete proposal on how Webpack integration might be setup at a high level. (Although really the only thing the user should have to do is ad the --use-webpack flag and everything else should _feel_ or _work_ the same.

@Brocco @hansl we can communicate about this more at length but I'd not only like to hear your thoughts, but community thoughts and opinions as well. Doing this integration _the right way_ should pave the way for giving people the option to have **other integrations** _the right way too_.

UPDATE: to add requirements.

#### Requirements 
- [ ] existing preprocessor support
  - [x] less
  - [x] sass
  - [ ] compass (?)
  - [x] stylus
  - [x] postcss (with plugins?)
  - [x] style sourcemaps
- [x] tree-shaking and dead code elimination
  - [x] TS2.0? es6 => es5
  - [x] Babel?
  - [ ] tsickle+gcc@advanced?
- [x] cacheable vendors chunk w/ cache busted main app
- [x] fast build and serve
  - [x] incremental build and in memory fs cache
  - [ ] multicore/cluster/forked worker support
- [x] support for existing projects
  - [x] migration path (if needed)
  - [x] test against test-app
  - [x] test against material 2
- [x] 3rd party module support
  - [x] should be as simple as `npm install` w/ import in app.
  - [x] CommonJS, CommonJS2, AMD, ES6, UMD
- [x] testing support
  - [x] karma integration
    - [x] configurable karma.conf yet bundle 'protected'
    - [x] w/ Istanbul coverage
  - [x] e2e w/ protractor
- [x] JS sourcemaps support
  - [x] pre/post trans?
- [x] dynamic on demand bundles
  - [x] `System.import` support which creates async dynamic bundle
- [ ] Routing 
  - [ ] Async Route Handling automagically?
- [ ] ngc support
  - [ ] if not ts.LSH support, build step
  - [ ] should not be slow
- [ ] asset support
  - [x] assets copied over in distribution bundle
  - [x] automagically resolve images
    - [ ] if under x size, base64 and inline image
    - [ ] if over x size, serve asset inside of assets folder
    - [ ] support for fonts
    - [ ] inline SVGs?

**Comment by SethDavenport** [comment]
+ a million.  Jokes aside I'm down to help

**Comment by gdi2290** [comment]
![starwars-star-wars-the-force-awakens-bb-8-3o7abB06u9bNzA8lu8](https://camo.githubusercontent.com/ad404dacefc8214c9b527acc2e1b483ccf5ee7b1/68747470733a2f2f6d65646961342e67697068792e636f6d2f6d656469612f336f3761624230367539624e7a41386c75382f67697068792e676966>)

**Comment by nickroberts** [comment]
Love. Definitely would like to have the option to use Webpack. @TheLarkInn's talk at ng-conf really got my juices flowing. I'm just starting to learn Webpack in more depth, and it seems like a great option.

So far, the SystemJS module loader works great, but seems a bit large (and a little slow) for development purposes.

That being said, I'm really excited about the offline template compiler using Rollup, based on what @robwormald demoed at ng-conf.

Exciting times!

**Comment by nickroberts** [comment]
Also, I was thinking that this functionality could probably be separated out to a plugin (or addon, as Ember calls it). This plugin could house all of the blueprints that would be needed to get things running.

You can already generate an addon via `ng addon <addon-name>`. I'm not entirely sure how useful it is, at least for angular-cli yet, though.

Perhaps the `ng new myApp --webpack` option could automatically check for this new addon, and either use it, or fail and provide instructions on how to install it.

Just some thoughts...

**Comment by mattdistefano** [comment]
FWIW I don't care about webpack integration per se, but parity with webpack's feature set would be nice to see. I can't claim much expertise here, but my understanding is SystemJS is actually superior in a couple aspects, namely that it is using a standards(ish)-based module format and loader runtime, and provides an easy path forward to "unbundle" for http2. Webpack 2 may address some of that, but fundamentally it seems like SystemJS is a module loader first, and bundler second, whereas Webpack is a bundler first, and loader second.

That said - I'm coming from a webpack background, and, trying out the CLI a week or two ago, I found myself missing the non-JS loading/dependency tracing that webpack does (i.e. being able to import a SCSS file and knowing that webpack will compile it and add any referenced assets - images, fonts, etc - to my dist output), along with the production build file name hashing. I think SystemJS itself actually supports both features, so this may just be a documentation issue?

Likewise, I logged #893 about this already, but I think HMR is going to be very important (as in, important enough that if webpack supports it and CLI doesn't, lots of devs are going to stick with webpack).

**Comment by SethDavenport** [comment]
My point of view is that I want a bundler, not a module loader. A module loader might makes sense once HTTP2 is mainstream but I'll believe in HTTP2 when I actually see it.  In HTTP1 module loaders are a solution looking for a problem IMO.

The non-JS bundling is actually my favourite feature of webpack. Suddenly with chainable loaders and PostCSS we have a really nice CSS workflow - on par with Babel.

**Comment by mattdistefano** [comment]
Right, but if you could have dynamic module loading and bundling today + painless unbundling for HTTP2 tomorrow, that would be the ideal scenario, yeah? And from my understanding of it, that's basically what SystemJS does, albeit with maybe some more configuration overhead than webpack. IIRC the CLI is already using the bundling features for prod builds too.

More broadly, I think the point of the CLI, to at least *some* extent, is to abstract the details of the build tooling, test runners, etc. So that's why I say I don't care about webpack vs. SystemJS vs. whatever - if the CLI can incorporate the useful features of webpack without incorporating webpack itself, that's fine with me.

**Comment by Foxandxss** [comment]
The problem I see @mattdistefano is that even if the CLI abstract you from System.js / webpack / foo. You still need to configure extra tools for your own needs. There you can't avoid to touch System.js / gulp / webpack.

**Comment by mattdistefano** [comment]
@Foxandxss Yeah, I'd still expect to have to tweak it to some extent. I guess my point is just that "webpack has some features the CLI lacks today" doesn't necessarily lead to "we must support webpack".

**Comment by TheLarkInn** [comment]
@mattdistefano Mind you I did say "Webpack and friends". I just believe the most popular bundler and build tool would be a great one to start with.

**Comment by SanderElias** [comment]
Point is, that in the current situation (read http2 not widely available) bundling is not an option. Even in development it's hurting. Compare an systemJS hello-world with an webpack one. The waterfall in the systemJs case does really hurt. Even more if you need to work on a networked connection!
(systemJS 293 resources vs 4 to 5 for webpack) 
So, a webpack option for the CLI is something we really need. 

Then there is the point that, with so many developers, opinions differ. Are we really in a position that we want to exclude a lot of them, simply because they don't like systemJS? 

And while we are on this subject of integrating another loader/bundler/packer/whatever, perhaps it's an idea to cater for future expansions in this area. For example RollupJS is also approaching. And I highly doubt that will be the last invention.

**Comment by mattdistefano** [comment]
Yeah but, again, there may be ways to solve those problems within the existing CLI tooling. That's what I would like to understand better - how many of these issues can be attributed to fundamental limitations of the CLI's current tech, and how many are really just features that haven't been implemented yet, or bugs that can be fixed?

**Comment by TheLarkInn** [comment]
If we implement this _ the right way_ we should be able to add system bundler as an option also in the future.

**Comment by mattdistefano** [comment]
I guess I'm still hung up on why there needs to be an _option_ at all. I'm not against webpack. I use it in production today and enjoy it and it's my team's choice for angular 1 and 2 apps in the near term.

But, for me, a big appeal of the CLI (and angular2 in general) is that it **is** opinionated. About project structure, architecture, tooling, etc. So when we start saying that, you know, this and that aspect of the CLI's internals need to be configurable so you can use whatever tooling you want, that's a big red flag for me. Because it means that I can't just sit down and use the CLI - I first have to learn about and choose between various bits of build system minutiae. It means that, when I have an issue, I can't just go find a solution - I'll have to wade through dozens of blog posts and SO threads that are almost but not quite useful because they address toolchain *x* while I opted for toolchain *y*. And I can only assume it complicates development of the CLI itself quite a bit. 

So, again, for me, if these issues can be addressed _without_ introducing that sort of fragmentation into the CLI, that would be the better option.

**Comment by kylecordes** [comment]
I am very excited about the possibility of CLI reaching a level of completion and polish that it "just works" for the great majority of all cases. Where even someone without much knowledge of these things (knowledge I have spent many months absorbing, and would prefer to forget) can get both an efficient development cycle and efficient production build. Where this issue tracker starts to get boring because it all "just works".

I think it will be most wise to achieve that before digging into swapping out the underlying toolset (always or optionally) with webpack. Webpack a great tool, just in terms of enabling mass, fast A2 adoption, we need one path that works great, more than N paths in various states of completion.

**Comment by SanderElias** [comment]
@kylecordes, While I agree with a large part of your sentiment, I stil believe supporting webpack asap is important to the CLI project.

While SystemJS is a great tool, it works very poorly if you have a slow connection to your storage(hdd). There are a lot of developers that need to work over an networked connection. In this situation, a save&reload cycle can easily take up over 10 seconds.  Webpack, in the same situation does the same in under 2 seconds. 
This alone is enough reason to support webpack, aside from all the other benefits.

For the end user of NG-CLI it's (or at least should be)  a one-time setting. Well, until the time they are ready to deploy that is.  

Bottom line is, if the CLI is not supporting webpack, it is in it's current state almost unusable for a large group of developers. I don't think that's the idea for the official angular toolkit....

**Comment by kylecordes** [comment]
@SanderElias I agree completely that performance is a big deal. I will think about this some more and try to suggest some ideas on how to improve it. I have found in our work here that webpack is also not the most performant way to work during development, but I have more to say about it then can reasonably go in this somewhat related issue discussion.

**Comment by mattdistefano** [comment]
Maybe I'm missing something, but given that SystemJS _can_ bundle, would it not be simpler to expose an option to use bundled output for development too? For example, I just commented out [the bundling's production environment check](https://github.com/angular/angular-cli/blob/67b577deb3bbee27228ec0b559bca6ba90603dcc/lib/broccoli/angular2-app.js#L158) and got bundled output with `ng serve`. I'm not saying that's necessarily the fix (perhaps bundling isn't performant at scale, perhaps sourcemaps get messed up, obviously we're still missing non-JS bundling support, etc) - point is just that there's _potentially_ a solution to be found within the existing tooling.

**Comment by mattdistefano** [comment]
I should add, HMR (which IIRC SystemJS, or at least JSPM, already supports) would also solve a lot of these issues.

**Comment by kylecordes** [comment]
@SanderElias and others -  here is the essence of what I was hinting at above. I think you will find it the currently most performant way to compile a *lot* of typescript code and deliver it to the browser with minimum filesystem activity: let TSC perform the application code bundling, and don't re-bundle libraries except when they really change. This results in far fewer roundtrips to the disk than (for example) the current CLI (broccoli). I implemented this as a fork to the quickstart for a quick demonstration:

https://github.com/kylecordes/quickstart/commit/3cf3041dd80ee8701a3cffb0cbe6264784017a2b

I know of others using this technique at scale in addition to us. However, I have no idea if something like this is suitable for the CLI offhand.

**Comment by SanderElias** [comment]
@kylecordes, @mattdistefano, I'm not arguing against SystemJS. SystemJS is great. That is not the point. Oh, and I'm pretty convinced that all problems eventually can be solved with it. Still not the point.

I highly agree with @TheLarkInn that when implemented the right way, a bundler/loader should be swappable. We talked about webpack. But it's not just webpack, there are others. An each one has it's own primary purpose. Heck, it should be even possible to use a different one of production. Production is a whole other discussion, but I would like to have the option to use for example tsickle and closure for that. 

Assuming that systemjS is the right tool for all projects that will be build with Angular 2 is not the way to go. As far as I'm aware angular-cli is as inclusive as angular itself. That also brings a responsibility. And this project is aimed to be the CLI for all angular things right? So the responsible thing to do is to cater for that. Not the easiest thing to do, but to me, it is the right thing to do.

**Comment by mattdistefano** [comment]
@SanderElias Well, I'd say being able to solve these problems within SystemJS *is* the point, at least to the extent that webpack has been advanced in this thread *as the solution* to these problems. If you think, philosophically, the project should support different tooling, that's fine. But that's a different argument than "we need webpack to have bundling".

In any case, on a philosophical level, I'm not wholly opposed to the CLI supporting other tooling, but I am concerned that a) we're going to get a bunch of half-baked integrations with [JS build tool of the week] rather than one really good integration with SystemJS and b) all these new integrations will make getting started and finding help far more difficult. To me, the CLI's first responsibility is to produce a cohesive, functional, and well-documented development experience that makes it trivial to get started using advanced angular concepts while enforcing best practices and providing a degree of consistency between projects. And I suspect that, the more user-configurable the build tooling becomes, the less achievable that goal is.

**Comment by hansl** [comment]
Here's the official stance of the CLI;

1. Saying "the elephant in the room [...] is webpack" is naive at best. You're talking about a technology, not a feature. The point of the CLI is that the user shouldn't have to worry about technological choices, only the features it (the CLI) provides. Whether it uses webpack, broccoli, or custom made technologies is irrelevant to the user.
1. A list of features as requirements for the build system (see below) has been fleshed out. These are considered blockers for the stable RC0.
1. There's a Proof Of Concept that's being built right now. Once and if we decide to go down the webpack road, your involvement will be welcome. If we go with SystemJS, your involvement is also welcome, of course.

The features we are looking for in a build system (before considering ourselves stable):

1. Support for sourcemaps. Shouldn't be hard, really. SystemJS already provides that.
1. Vendor should be bundled. Don’t download and/or rebuild vendor files, unless necessary.
1. Project Code can be unbundled. Not a requirement, but makes it easy to debug when SourceMaps are a problem.
1. Better third party support:
    a. Node Package Awareness. Using `package.json`, figure out the `vendorNpmFiles` and the path mapping automatically, without developer intervention.
    1. Have a better story where npm install is the only thing needed to use some third party libraries.
1. Offline Template Compiler  (might be optional for local dev, required for production)
1. Tree shaking.
1. Assets Moved Smartly (css/html copied). Everything outside of `public/` should be copied if needed (templates/styles), compiled if needed (sass/less/... and jade/html/...), inlined if needed. Everything inside of `public/` should be copied to `dist/` folder, compiled if needed.

**Comment by SanderElias** [comment]
Thank you for setting us straight @hansl!

I do agree with all what you said. What triggered this discussion is not actually the missing of webpack, but rather missing some of the points in your feature list.

However, there is something missing in this feature list. And it's not something easily achieved.
namely **developer performance**. To be more precise, the time the tooling needs to display a saved change to a browser. And as a close second, the time it takes to run the tests.

Let me explain, while everything may be hunky-dory on your local SSD *nix Intel I7 driven system, there are large number of developers that not have that luxury. 
There are a lot of places, where developers need(are forced...) to use a networked drive, or can (not yet) afford to run on the latest hardware. 
There are plenty of reasons why not everything is on a local SSD. All of them are valid, and more important: out of our control.

**Comment by kylecordes** [comment]
@SanderElias  I think your observation about performance is very important. I hope performance makes it onto the list that @hansl wrote above, I think his seven items should include your eighth item.  This will be important for adoption.

But I'm not sure your focus on webpack vs System really gets to the heart of the problem. System is not the source of the slowness you are seeing. Rather, what you are experiencing I believe has to do with broccoli. This is a much discussed issue with broccoli, broccoli does extensive local disk temp file manipulation, creating thousands of files (mostly symlinks) when used in the typical manner on  a small project, and many more a large project. I don't know if this is a solvable problem with broccoli. I don't know how wedded the angular-cli project is to broccoli. I do know that broccoli is conceptually brilliant (shout out to @joliss), in spite of this performance struggle on Windows. (It's really not about SSD, is about filesystem optimization and performance. Mac and OSX can manipulate thousands of small local files very  quickly on a local file system. Windows, not so much, it is optimized in other directions. Across a remote file system, manipulating thousands of files is always slow regardless of platform or hardware.)

Going back to webpack, it is not a panacea. I have found that it also becomes awkwardly slow as project size grows. Even on a fast local SSD on a MacBook Pro.

Experimentally I found that the very fastest development cycle occurs when you bundle dependencies into a small number of bundles, **only** when they change (and not touch your library dependencies at all after each application code edit). Then it use the ability of typescript to emit a system bundle, And simply forgo all of the other bundling tools for application code. TypeScript can do this very quickly if you just let it, and produce a file immediately ready to consume with SystemJS. If instead you have typescript emit numerous JavaScript files which then must be processed by a downstream tool, things are never as fast. I have confirmed with other development teams who have independently reached the same conclusion, that this is the fastest way to set up development tooling on a large Angular 2 project at the moment. I wrote about it here:

http://kylecordes.com/2016/fast-angular-2-compile-cycle-with-tsc

I have no idea whether this approach might be within scope of contemplation for angular-cli, and I am very excited about CLI (because I think it will help considerably with Angular 2 adoption) even if it does not aim for this best available performance. But I do think it would be good if people on the CLI team would at least be aware of this fastest available path, and have it out there is a conceptual target for somewhere down the road.

**Comment by SanderElias** [comment]
@kylecordes My points were actually never really about webpack. I mentioned webpack because it was an solution to a problem, as I saw it at that moment. Being in this discussion, it forced me to dig in a bit deeper, and now I know a tad more ;)

Frankly I don't care how ng-cli works, nor what tools it uses internally, as long as it hits every mark on the feature list (**and in my viewpoint the feature list should include the developer performance**). 

The main point is, the ng-cli should be the 'go-to' tool for everything angular. If it hits everything on the list but skips the performance issue, it still is a no-go for a lot of developers. And that is very sad. Angular 2 is great, but I suspect a lot of (beginning) devs will be put off by the tooling needed. To be fair, even for versed developers the tool-chain needed now is challenging. 

Bottom line, **if the performance off the angular-cli is sub-par, this is bad for angular 2 acceptance!**

**Comment by robwormald** [comment]
@hansl we should probably add "lazy loading" to that list of reqs, or at least "doesn't preclude programatic lazy loading in a sane fashion"

Webpack2 is better about this, I'm just not sure on the status of where it is in beta land

**Comment by scottwio** [comment]
+1 for webpack for 3 reasons. 

1) Including npm packages becomes really simple just import the file you want, rather than the current solution https://github.com/angular/angular-cli/wiki/3rd-party-libs even if angular CLI can automate that process it means to manually change an import entry point you have to understand the inner workings of systemJs and the Angular CLI.

2) It's already used in the community. This may seem like a small thing but it means that the javascript community can build new plugins together and share knowledge / code. A lot of developers already know how to use webpack so it's not "another thing to learn" like systemjs and broccoli.

3) Loading all files and dependencies in the browser is slow. From my experience bundling speeds everything up. Bundling is not a silver bullet but it's far better than loading in the browser.

**Comment by SethDavenport** [comment]
For me, the argument for Webpack is that it's so widely used. My (possibly flawed) perception is that it's as close to a 'defacto' as anything in the greater JS community right now.

So when I look into the CLI and I'm like: 'what is this broccoli thing', it seems that 'why not Webpack' is a fairly natural question.

As for SystemJS, I'm familiar with it, I just don't like it :). It's too config-heavy. Webpack is what saved me from that particular merry-go-round.

That said, it's quite possible that I'm living in a weird bubble where everyone I typically collaborate with uses Webpack (I came to Angular 2 via React after all). But it would be soo nice not to diverge the common React and Angular toolchains even more than they already are.

**Comment by un33k** [comment]
@SethDavenport I like the ng2-redux-starter. What are the chances to have video tutorial on usage?

**Comment by joliss** [comment]
On the issue of Broccoli being slow on Windows: There have been discussions about avoiding writing out files (or symlinks) and instead passing "virtual" in-memory files between plugins. I don't think this has gotten very far yet, but I'm definitely sympathetic to the idea.

On "Webpack vs. Broccoli", I think the two don't try to solve the same problem, so it's not that meaningful a question. Yehuda [said it well](https://twitter.com/wycats/status/738157997562597377):

> I think it's best to see broccoli as a primitive that you could use to build something like webpack or ember-cli

**Comment by kushwahashiv** [comment]
I'm a big fan of webpack. We have recently wrote an allocation and it supports is awesome.

+1

**Comment by SethDavenport** [comment]
I understand they are not equivalent. Broccoli requires several other tools that together provide a complete solution. Webpack is a complete solution by itself. That's largely my point.

**Comment by SethDavenport** [comment]
Why rebuild the system out of primitives when there's already a solution in wide use in the community?

**Comment by scottwio** [comment]
I spent sometime this weekend building mini projects with both the current CLI and the Webpack tutorial found on the angular.io website. 

It really is night a day when it comes to the development experience. Webpack feels solid and compiles quickly. The system.js solution is confusing, I spent a lot of time googling it's quirks rather than developing. Bring all the files together on each reload in browser is also painfully slow even on the smallest project.

**Comment by robwormald** [comment]
@sethdavenport webpack is a complete solution for *bundling*, and it does a very good job at that. It doesn't, however, provide any support on testing, code generation / blueprints, etc. So while our team likes the idea of webpack for bundling, it's not a complete solution to the problem. 

I'm pretty excited about webpack2, but it should be noted that it uses SystemJS for lazy loading (which is imho, absolutely the right idea as it tracks the future of the web platform and the Loader spec). This is part of the issue with System, since it has to run on the client side it knows nothing of folder structure or node's module resolution algorithm. 

My concern is that we trade one set of configuration (system config) for another (webpack config) and end up not providing an end to end experience. Any solution that relies 100% on webpack is going to be incomplete, so other tools will be involved. 

I'm of the opinion that offloading the (difficult indeed) problem of bundling to webpack will let us focus on the other stuff, but it's not quite as simple as "switch it all to webpack". Additionally, making it an opt in is a non-starter, as that means we end up maintaining two sets of plugins. So we have to evaluate carefully to be sure webpack + whatever will be a good solution.

**Comment by SethDavenport** [comment]
@robwormald - fair enough.  Turns out bundling is typically the only problem I need help with - testing etc.is easy enough calling special purpose tools like mocha from npm's `scripts`.

Asynchronous module loading is a cool idea in theory but iis one of those things that I've never actually needed in a real app. The bottlenecks have always been somewhere else. Maybe one day I'll need it - who knows 😃

Finally for the codegen stuff - I guess I need to wait to see it in action before I understand the value. It's not really in my feature list as a user of the framework right now.

**Comment by shlomiassaf** [comment]
We are working with both webpack and angular-cli on the same project.

Not everything is perfect but we are getting there.
Webpack configuration is based on angular2-webpack-starter.

We use angular-cli for code generation.
We use webpack for testing (unit & e2e) via npm script.
We are working on testing to work with angular-cli (we have a config_angular_cli directory for all config files)

We will also get the build running.

We had to found some common ground, which are usually around file structure.
We chose the angular-cli file structure where e2e files are in the src root (in webpack-starter it's different)

The only issue we can't solve right now is `index.html`, as a workaround we have 2, one for each setup, and our script just copies the right one at the end. 
This means that dev environment can't run since we need `index.html` so webpack was chose for development dev server. (currently don't run the cli's dev server).

I think with something like rewrite url we can make it work for the cli as well.

**Comment by TheLarkInn** [comment]
@shlomiassaf You can use some of the more advanced features of html-webpack-plugin on an ENV basis. But good to note that you are coming across this challenge using the CLI with webpack.

**Comment by ribsies** [comment]
Hello, first time poster here but I just want to echo what everyone is saying about webpack.

I don't think the main draw of angular-cli is that it's opinionated, i think the draw is that it's very easy to get up and running. After spending a good amount of time working with angular CLI I found it lacking in a couple fairly large areas.

One being the production builds not bundling or working. This was a big one

Also the relative difficulty of adding third party libraries.

Over the past week I've been messing around with webpack and found it extremely easy to use for development and production. Tests were also really easy to get up and running.

After that experience I decided to change my clients project over to webpack. I gutted pretty much everything the CLI generated and replaced it with the much simpler webpack.

Now all our builds work flawlessly and as a bonus our HTML and CSS are all compiled in line! That's pretty huge. Now all we use the CLI for is generating files.

Currently I am working on replacing that as well so I can get the requires in for my components when I generate scripts. That would make my life easier.

My point is, I want this and support this idea of having webpack as an option for the CLI.

**Comment by kushwahashiv** [comment]
@ribsies I love to use webpack in client development and use the same exactly as you use. I generate routes ect using webpack and do little modification to it to make it compatible to my structure.
webpack is really awesome for various works testing, development, prod deployment configuring a lot other stuff environment specific etc.

**Comment by johnpapa** [comment]
I like the idea of options ... but we may have other options later too. so instead of a systemjs or webpack option (and later foo and bar options), we could add a `--build` flag (or build type or whatever its name is)

perhaps a flag like `ng new --build=webpack`

**Comment by mattdistefano** [comment]
FWIW I'd much rather see the CLI support *one* build system extremely well and offer a code-generation-only option for those who prefer to use something else and manage its config manually.

**Comment by TheLarkInn** [comment]
#### Requirements 
- [ ] existing preprocessor support
  - [ ] less
  - [ ] sass
  - [ ] compass (?)
  - [ ] stylus
  - [ ] postcss (with plugins?)
  - [ ] style sourcemaps
- [ ] testing support
  - [ ] karma integration
    - [ ] configurable karma.conf yet bundle 'protected'
    - [ ] w/ Istanbul coverage
  - [ ] e2e w/ protractor
- [ ] tree-shaking and dead code elimination
  - [ ] TS2.0? es6 => es5
  - [ ] Babel?
  - [ ] tsickle+gcc@advanced?
- [ ] cacheable vendors chunk w/ cache busted main app
- [ ] asset support
  - [ ] assets copied over in distribution bundle
  - [ ] automagically resolve images
    - [ ] if under x size, base64 and inline image
    - [ ] if over x size, serve asset inside of assets folder 
    - [ ] support for fonts 
    - [ ] inline SVGs? 
- [ ] fast build and serve
  - [ ] incremental build and in memory fs cache
  - [ ] multicore/cluster/forked worker support
- [ ] support for existing projects
  - [ ] migration path (if needed)
  - [ ] test against test-app
  - [ ] test against material 2
- [ ] dynamic on demand bundles
  - [ ] `System.import` support which creates async dynamic bundle
- [ ] 3rd party module support
  - [ ] should be as simple as `npm install` w/ import in app.
  - [ ] CommonJS, CommonJS2, AMD, ES6, UMD
- [ ] ngc support
  - [ ] if not ts.LSH support, build step
  - [ ] should not be slow
- [ ] JS sourcemaps support
  - [ ] pre/post trans?

**Comment by shlomiassaf** [comment]
Integration with `HtmlWebpackPlugin` ?

**Comment by shlomiassaf** [comment]
BTW, @TheLarkInn , When I think about it, I see NG CLI as a service... something like the typescript language service...

It has a top level structure that knows everything about the application, from there moving from this to that (webpack to SystemJS) is just a matter of transformation.

It's complicated, I know... it depends on the ember CLI I know... just a thought

**Comment by TheLarkInn** [comment]
@shlomiassaf html templating will be supported.

**Comment by shlomiassaf** [comment]
@TheLarkInn Will the user be able to use the `HtmlWebpackPlugin` as usual, through the config?

Bundles/Chunks with hash requires a template form on index.html

**Comment by TheLarkInn** [comment]
> Whether it uses webpack, broccoli, or custom made technologies is irrelevant to the user.

Keep in mind what @hansl mentioned above when referencing:
https://github.com/TheLarkInn/angular-cli/blob/webpack/addon/ng2/models/webpack-build-config.ts

**Comment by shlomiassaf** [comment]
Abstract the build system, configuration is done in an ng-cli fashion...

**Comment by TheLarkInn** [comment]
I believe that there's merit to it, however it's something that is a 'nice-to-have'. Advanced webpack users are already going to be opinionated on their setup so maybe the CLI is not for them.

I do see the merit in it, but I'm not going to personally going to invest the time _yet_. Maybe it's something _you_ can look into integrating off my branch when things near completion.

**Comment by shlomiassaf** [comment]
I totally agree that first a stable integration is required 1st.
Once this works we can explore advanced scenarios.

BTW, since CLI gives offline compilation its not an option not to use it...

**Comment by vapits** [comment]
I have to agree with @SanderElias, CLI can really affect the way people seeing Angular 2. In my team a colleague shared with me his thoughts - after "fighting" a little bit with SystemJs trying to figure out a way not to have 600+ requests for a simple app client side - about Angular 2, and those wasn't good at all.

It was disappointing for me a teammate being not happy with a decision we made and even that is not Angular's fault, the bad reputation goes to Angular 2.

I do understand the huge demand for features, huge anticipation for the CLI to support a huge amount of solutions but being still in early beta state I think it's not the time to provide everything.

Personally I would prefer Webpack for several reasons (mostly production related) but as @scottwio said if you spend (and I know you all spend lots and lots of hours) to use CLI you'll see that is not the case right now.

Of course a complete solution will be great (like choosing what to use for your project as more solution will be available in the near future) but a more stable CLI and a thought about a realistic production approach even with SystemJs is more than enough. Sometimes reading the codes I feel that even a bundler will solve like the 50% of the complains.

I hope everything make sense :P
Keep up the GREAT work guys!

**Comment by intellix** [comment]
Currently without bundling, HMR or http/2, every file change makes a request for 650+ files here and it takes about 7 seconds before seeing changes, which is extremely painful during development. I don't experience that with the webpack starter cause of the bundling.

**Comment by kushwahashiv** [comment]
@intellix  exactly the one of reasons we pick webpack.

**Comment by un33k** [comment]
Not a frontend guy yet, but I am very interested thanks to NG2.  Lot's of back & forth between webpack and ng-cli and can't decide which way to go.  Found [this article](https://medium.com/@martin_hotell/angular-cli-and-real-development-productivity-4ba3c1865eda#.mbuubbaqa) which explained everything in simple terms and made it easy for me to lean towards webpack.  NG-CLI is a great concept, but with little exposure so far I found my hands tied behind my back everytime I wanted to do something a bit different.  The experienced frontend folks are very opinionated when it comes to their tools & environments and rightfully so.  They are also the one that can bend the tools to their wills.  On the other hand, inexperienced folks like myself are at the mercy of the tools and (IMHO) ng-cli at this stage of development with the current mindset trajectory, might miss all of `us`.

**Comment by iloveyo123u1** [comment]
+1

**Comment by mmrath** [comment]
I started with cli and now when the project has grown a bit, I started to realise the slowness during development. I now make change and save got to the browser after a few seconds. I changed the project to use webpack starter it is much better. Apart from development performance another thing that is a pain with the current cli is windows support. I think the cli will not be usable by mass corporate users if it does not have good windows support.

**Comment by dancancro** [comment]
I made a website out of information I've collected for the past few years. It is meant to help people choose between things based on the benefits they deliver. Maybe it will help with this discussion. Let me know if you see any errors or enter them into the forms on the page.
http://www.techtradeoffs.info
